src/rules/iddataweb-verification-workflow.js

```diff
@@ -0,0 +1,140 @@
+ * @category marketplace
+ *
+ * This configuration allows you to add an ID DataWeb verification workflow to the user’s first time login experience. This will allow your organization to ensure the user is really who they claim to be, aligning with KYC/AML requirements, and mitigating fraud attempts.
+ *
+ */
+
+function iddatawebVerificationWorkflow(user, context, callback) {
+  const { Auth0RedirectRuleUtilities } = require("@auth0/rule-utilities@0.1.0");
+
+  const ruleUtils = new Auth0RedirectRuleUtilities(
+    user,
+    context,
+    configuration
+  );
+
+  const {
+    IDDATAWEB_BASE_URL,
+    IDDATAWEB_CLIENT_ID,
+    IDDATAWEB_CLIENT_SECRET,
+    IDDATAWEB_LOG_JWT,
+    IDDATAWEB_ALWAYS_VERIFY,
+    IDDATAWEB_PREFILL_ATTRIBUTES,
+  } = configuration;
+
+  const idwTokenEndpoint = `${IDDATAWEB_BASE_URL}/axn/oauth2/token`;
+  const idwAuthorizeEndpoint = `${IDDATAWEB_BASE_URL}/axn/oauth2/authorize`;
+  const auth0ContinueUrl = `https://${context.request.hostname}/continue`;
+
+  // initialize app metadata
+  user.app_metadata = user.app_metadata || {};
+  user.app_metadata.iddataweb = user.app_metadata.iddataweb || {};
+
+  // if the user is already verified and we don't need to check, exit
+  if (
+    user.app_metadata.iddataweb.verificationResult === "verified" &&
+    IDDATAWEB_ALWAYS_VERIFY === "off"
+  ) {
+    console.log("user " + user.user_id + " has been previously verified.");
+    return callback(null, user, context);
+  }
+
+  // if coming back from redirect - get token, make policy decision, and update user metadata.
+  if (ruleUtils.isRedirectCallback) {
+    console.log("code from IDW: " + ruleUtils.queryParams.code);
+
+    let options = {
+      method: "POST",
+      url: idwTokenEndpoint,
+      headers: {
+        "Cache-Control": "no-cache",
+        Authorization:
+          "Basic " +
+          Buffer.from(
+            IDDATAWEB_CLIENT_ID + ":" + IDDATAWEB_CLIENT_SECRET
+          ).toString("base64"),
+        "Content-Type": "application/x-www-form-urlencoded",
+      },
+      form: {
+        grant_type: "authorization_code",
+        code: ruleUtils.queryParams.code,
+        redirect_uri: auth0ContinueUrl,
+      },
+    };
+
+    request(options, function (error, response, body) {
+      if (error) {
+        return callback(err);
+      }
+
+      let decodedToken;
+      try {
+        decodedToken = jwt.decode(JSON.parse(body).id_token);
+      } catch (tokenError) {
+        return callback(tokenError);
+      }
+
+      console.log("policy decision: " + decodedToken.policyDecision);
+      console.log("score: " + decodedToken.idwTrustScore);
+
+      if (IDDATAWEB_LOG_JWT === "on") {
+        console.log(JSON.stringify(decodedToken));
+      }
+
+      // if IDW verification was successful - update user's metadata in Auth0.
+      //this could be used for downstream application authorization,
+      //or mapping access to levels of assurance.
+      if (decodedToken.policyDecision === "approve") {
```

Comment:
Just to confirm ... this means that anything other than "approved" would leave no record in Auth0 that IDV ever happened and, in this case, would respond to the Client with nothing different in the identity. Is that how it should work? Might be good to mention that in the integration guide.

Reply:
I will rework this to include the policy decision and timestamp.